---
title: "Accounting for batch effects in [sleuth](http://pachterlab.github.io/sleuth/about)"
author:
- Harold Pimentel
- Nicolas Bray
- Páll Melsted
- Lior Pachter

output:
  html_document:
    toc: false
    theme: default
    fig_height: 10
    fig_width: 10


abstract: "This walkthrough teaches the use of sleuth for analysis of experimental designs with multiple covariates via an example of conditioning on batch effect."
---

# Warning

This tutorial is outdated with the release of sleuth v.0.30.0, which performs gene differential expression by aggregating p-values from transcript differential expression. This tutorial will be retained here for purposes of maintaining a record but will no longer be supported. We recommend that users switch to the updated version of sleuth.


# Introduction
NOTE: This walkthrough is outdated with the release of sleuth v0.30.0, which uses transcript p-value aggregation instead of transcript count aggregation for performing gene differential expression. A new version of this walkthrough corresponding to sleuth v0.30.0 is available here: [insert HML]

This walkthrough teaches how to test for differential expression in a way that is informed by known batch effects in the data.
The example uses data from the paper Bottomly et al., [Evaluating Gene Expression in C57BL/6J and DBA/2J Mouse Striatum Using RNA-Seq and Microarrays](TODO), _PLoS One_, 2011.
In addition to this walkthrough, data from the paper was is used for some of the [analyses in the sleuth paper](TODO).
The specific data that are examined in this walkthrough are 21 RNA-Seq samples from striatal samples extracted from two different mouse strains.
The strains are C57BL/6J (B6) and DBA/2J (D2).
The Bottomly _et al_. analysis focuses on a comparison of the RNA-Seq gene expression estimates to microarray estimates.
In this walkthrough we examine in detail how to analyze the RNA-Seq.

The walkthrough is self-contained, but new users to sleuth will benefit from first studying the Trapnell _et al_. walkthrough which demonstrates the use of sleuth on a simple two condition experiment.
This walkthrough, in addition to illustrating how to condition on batch effects, provides an example for how to use sleuth for analyses of experimental designs with multiple covariates.
The tutorial also moves beyond the isoform-level analysis of the Trapnell _et al_. walkthrough and illustrates how to perform gene-level analysis.

# Preliminaries

Create a directory for the analysis and start up `R`.
Set the working directory with the `setwd()` command.

Requirements for this walkthrough:

- knowledge of [how to use sleuth for simple two condition experiments](https://pachterlab.github.io/sleuth/walkthroughs/trapnell).
- `cowplot` for making prettier plots and plots with grids. Available in CRAN:  `install.packages('cowplot')`.
- `biomaRt` for extracting the Ensembl transcript to gene mapping. For alternative methods, [please see this walk-through](TODO).

To install the package:

```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite("biomaRt")
```

- The walthrough begins after the RNA-Seq samples have been quantified with kallisto. While kallisto is very fast in quantifying samples, downloading the raw data from the short read archive is time consuming. kallisto quantifications are therefore directly downloadable for the relevant samples using

```{bash eval=FALSE}
wget -O ../Bottomly_results.zip 'https://www.dropbox.com/s/gnbnkuup3pdfenk/Bottomly_results.zip?dl=1'
unzip ../Bottomly_results.zip -d ..
```

Let's load the requisite packages:

```{r}
suppressMessages({
  library('cowplot')
  library('sleuth')
})
```

## Parsing metadata

A sleuth analysis is dependent on a metadata file, which describes the experimental design, the sample names, conditions and covariates. The metadata file is external to sleuth, and must be prepared prior to analysis. The first step in a sleuth analysis is loading of the metadata file:


```{r}
metadata <- read.csv('../metadata/experiment.csv', stringsAsFactors = FALSE)
```

```{r}
head(metadata, n = 20)
```

This file describes the experimental design.
The first column lists the short read archive accession for each sample that is part of the experiment to be analayzed.
The biological condition is recorded in the "strain" column.
The last two columns describe how the sequencing was undertaken and reveal that certain batches were sequenced together.

```{r}
extract_metadata <- function(library_name) {
  ret <- lapply(strsplit(library_name, '_'),
    function(x) {
      data.frame(strain = x[1], experiment = x[2], lane = x[3],
        stringsAsFactors = FALSE)
    })
  dplyr::bind_rows(ret)
}

metadata <- dplyr::select(metadata, -strain)
metadata <- dplyr::bind_cols(metadata, extract_metadata(metadata$library_name))
metadata <- dplyr::select(metadata, run_accession, library_name, strain,
  experiment, lane)
```

Finally, we add the path names of the kallisto output directories to the metadata table:

```{r}
metadata <- dplyr::mutate(metadata,
  path = file.path('..', 'results', run_accession, 'kallisto'))
head(metadata)
```

It is important to spot check the metadata file again to make sure that the kallisto runs correspond to the accession numbers in the table, so that each row is associated with the correct sample.

```{r}
metadata <- dplyr::rename(metadata, sample = run_accession)
```

## Associating transcripts to genes

The sample quantifications performed by kallisto have produced transcript abundance and count estimates.
These have been parsed by sleuth in the steps just performed, however sleuth does not "know" about genes yet.
To perform gene-level analysis sleuth must parse a gene annotation.
One easy way to do this is using biomaRt and Ensembl:

```{r cache=TRUE}
mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "mmusculus_gene_ensembl",
  host = "dec2015.archive.ensembl.org")
  # host = "ensembl.org")
ttg <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "transcript_version",
  "ensembl_gene_id", "external_gene_name", "description",
  "transcript_biotype"),
  mart = mart)
ttg <- dplyr::rename(ttg, target_id = ensembl_transcript_id,
  ens_gene = ensembl_gene_id, ext_gene = external_gene_name)
head(ttg)
```

The resulting table contains Ensembl gene names (column 3) and the associated transcripts (column 1). Note that the gene-transcript mapping must be compatible with the transcriptome used with kallisto.
In other words, to use Ensembl transcript-gene associations kallisto was run using the Ensembl transcriptome.

# Preparing the analysis

The next step is to build a sleuth object. The sleuth object contains specification of the experimental design, the full model to be estimated from the data, a map describing grouping of transcripts into genes (or other groups), and a number of user specific parameters. In the example that follows, `metadata` is the experimental design, `~strain` specifies the full model to be estimated, and `target_mapping` describes the transcript groupings into genes previously constructed:

```{r cache=TRUE}
so <- sleuth_prep(metadata, target_mapping = ttg,
  aggregation_column = 'ens_gene', extra_bootstrap_summary = TRUE)
```

Notice the warning: "518 target_ids are missing annotations for the aggregation_column: ens_gene."
This means that there are transcripts that do not have a gene mapping.
This is pretty common in most annotations.
However, if this number is very high, it is advisable to consider using another transcript-to-gene mapping or another annotation.

# The naïve analysis

As discussed in the Trapnell _et al_. walkthough, to identify genes differential between two conditions two models must be fit by sleuth.
The first is known as the reduced model, and in this case we name the "intercept model".
It fits a single parameter for each gene thereby enforcing equal abundances between conditions.
The second model is known as the full model, and it includes a parameter that is strain dependent.
We call it the "strain model".
With the likelihood ratio test sleuth identifies genes whose abundances are significantly better explained when strain is taken into account. The mechanics of the fitting are:

```{r cache=TRUE}
so <- sleuth_fit(so, ~1, 'intercept')
so <- sleuth_fit(so, ~strain, 'strain')
```
The likelihood ratio test (lrt) is performed with

```{r}
so <- sleuth_lrt(so, 'intercept', 'strain')
```

Significantly differential genes are extracted with the commands

```{r}
sleuth_table_strain <- sleuth_results(so, 'intercept:strain', 'lrt', show_all = FALSE)
sleuth_significant_strain <- dplyr::filter(sleuth_table_strain, qval <= 0.05)
```

The most significantly differential genes are

```{r}
head(sleuth_significant_strain, 20)
```
To confirm that the differential analysis is working it's important to examine specific genes in detail. For example, the most significantly differential gene that emerges from the analysis above is [ENSMUSG00000066553](http://www.ensembl.org/Mus_musculus/Gene/Summary?g=ENSMUSG00000066553;r=12:24638669-24639010;t=ENSMUST00000140065)(a processed pseudogene), whose abundances in the different samples can be viewed with

```{r}
plot_bootstrap(so, "ENSMUSG00000066553",
  units = "scaled_reads_per_base", color_by = "strain")
```

The plot immediately reveals that the gene is expressed in B6 mice and is off in D2 mice. The boxes around the mean abundance estimate for each gene are computed from the kallisto bootstraps which are proxies for the variation that would be observed in technical replicates.

However visualization of individual gene abundances is not sufficient to fully quality control an experiment. A principal component analysis (PCA) of the kallisto abundances shows that while the samples appear to be clustered into a few distinct groups, they are not well-separated by strain.

```{r}
plot_pca(so, color_by = 'strain')
```

The source of the clustering is revealed by coloring the samples by "lane" instead:

```{r}
plot_pca(so, color_by = 'lane', text_labels = TRUE)
```

This batch effect is problematic, as genes might be identified as differential in the strain analysis merely because they were sequenced on different lanes.

# Batch analysis

To test whether this is in fact the case, the test described above can be repeated with the full model comprising a paramter for "lane" instead of "strain".

```{r cache=TRUE}
so <- sleuth_fit(so, ~lane, 'lane')
```

The likelihood ratio test is performed as before except with the lane model

```{r}
so <- sleuth_lrt(so, 'intercept', 'lane')
```

The differential analysis test now reveals genes most affected by batch:

```{r}
sleuth_table_lane <- sleuth_results(so, 'intercept:lane', 'lrt', show_all = FALSE)
sleuth_significant_lane <- dplyr::filter(sleuth_table_lane, qval <= 0.05)
head(sleuth_significant_lane, 20)
```

To identify genes that are truly differential between strain, it is therefore necessary to fit and account for the (lane) batch effect.

## Fitting a model that acounts for batch

To account for batch while testing for strain differences, the full model must include parameters for both lane and strain.

```{r cache=TRUE}
so <- sleuth_fit(so, ~lane + strain, 'full')
```

The reduced model must still include a parameter for the lane.
In this way, the strain differences are accounted for while conditioning on the lane.
This model, the "lane model" has already been fit.
So the likelihood ratio test can now be applied as before:

```{r}
so <- sleuth_lrt(so, 'lane', 'full')
sleuth_table <- sleuth_results(so, 'lane:full', 'lrt', show_all = FALSE)
sleuth_significant <- dplyr::filter(sleuth_table, qval <= 0.05)
head(sleuth_significant, 20)
```

While the top 8 genes are the same as in the naïve analysis, differences start to emerge afterwards.
Furthermore, the top genes emerge as more significantly differential when conditioning on batch.

## Comparing the results

This is a simple comparison just to show how the rankings differ.

First, we create a new column for the relative rank within both lists.

```{r}
tmp <- dplyr::inner_join(
  dplyr::mutate(
    dplyr::select(sleuth_significant, target_id, qval_batch = qval),
    rank_batch = 1:length(qval_batch)),
  dplyr::mutate(
    dplyr::select(sleuth_significant_strain, target_id, qval_strain = qval),
    rank_strain = 1:length(qval_strain))
  ,
  by = 'target_id')
```

Next, we compute the relative difference between the ranks.

```{r}
tmp <- dplyr::mutate(tmp,
  relative_difference = abs(rank_batch - rank_strain) / (rank_batch + rank_strain))
```

Finally, we can plot a few of the top different results:

```{r}
top_different <- dplyr::arrange(head(tmp, 100), desc(relative_difference))
p <- lapply(top_different$target_id[1:4],
  function(x) {
    plot_bootstrap(so, x, color_by = 'strain')
  })
plot_grid(plotlist = p)
```

## Comparing effect sizes

We can extract all of the parameters using the `extract_model` function by supplying the model name used in the `sleuth_fit` call.

```{r}
full_model <- extract_model(so, 'full')
strain_model <- extract_model(so, 'strain')
```

Next, if we wish to compare the values we can do so.
Note that these values are conditional on the other parameters being in the model.

```{r}
effect_sizes <- dplyr::inner_join(
  dplyr::select(dplyr::filter(full_model, grepl('strain', term)),
    target_id, full_estimate = estimate),
  dplyr::select(dplyr::filter(strain_model, grepl('strain', term)),
    target_id, strain_estimate = estimate),
  by = 'target_id'
  )
dplyr::inner_join(
  dplyr::select(top_different[1:10, ], target_id),
  effect_sizes,
  by = 'target_id')
```

There is not much to say other than they are different.
