---


title: "Differential analysis with multiple conditions and covariates"

date: "`r Sys.Date()`"
author:
- Harold Pimentel
- Nicolas Bray
- Páll Melsted
- Lior Pachter

output:
  html_document:
    toc: false
    theme: default
    fig_height: 10
    fig_width: 10

abstract: "This walkthrough teaches the use of sleuth for identifying differences in expression between any of a number of multiple conditions."
---

# Warning

This tutorial is outdated with the release of sleuth v.0.30.0, which performs gene differential expression by aggregating p-values from transcript differential expression. This tutorial will be retained here for purposes of maintaining a record but will no longer be supported. We recommend that users switch to the updated version of sleuth.

# Introduction

The purpose of this walkthrough is to demonstrate the use of [__sleuth__](http://pachterlab.github.io/sleuth/) for analaysis of an experiment in which it is of interest to identify differences that may exist between any of a number of experimental conditions. We illustrate this use case with data from the paper Boj _et al_., [Organoid models of human and mouse ductal pancreatic cancer](http://www.sciencedirect.com/science/article/pii/S009286741401592X), Cell, 2015.
One of the experiments the paper describes is RNA-Seq performed on syngenic mice organoids.
Specifically, RNA was extracted from murine normal (mN), pancreatic tissues that contained low-grade murine PanIN (PanIN) and pancreatic ductal organoids from multiple primary tumors (mT). The goal is to identify genes in _either_ of the PanIN or mT samples that differ in their expression from the control (nN) sample. The walkthrough explains how to setup and perform a test identifying such genes while taking into account the sex of each mouse.

# Preliminaries

Requirements for this tutorial:

- knowledge of [how to use sleuth for simple two condition experiments](https://pachterlab.github.io/sleuth/walkthroughs/trapnell).
- __kallisto__ quantified samples from GEO GSE63348 (SRA SRP049959).
- the short read archive (SRA) runtable that contains accessions for the samples analyzed and associates metadata to them. [Download here](https://raw.githubusercontent.com/pachterlab/sleuth_walkthroughs/master/boj/metadata/SraRunTable.txt).

To facilitate practice with the walkthrough we have made kallisto pseudoalignments for the samples available here:

```{bash eval=FALSE}
wget -O ../Boj_results.zip 'https://www.dropbox.com/s/j4bznighrbtj02i/Boj_results.zip?dl=1'
unzip ../Boj_results.zip -d ..
```

The SRA runtable looks like this:
```{r}
sample_to_condition <- read.table('../metadata/SraRunTable.txt', header = TRUE,
  sep = '\t')
head(sample_to_condition)
```

Some of the metadata that is not relevant and it is convenient to simplify the table

```{r}
sample_to_condition <- dplyr::select(sample_to_condition,
  sample = run, sex = Sex_s, genotype = mouse_genotype_s)
head(sample_to_condition)

```

Finally, __sleuth__ and cowplot (we like the default theme better) are loaded with

```{r}
suppressMessages({
  library('cowplot')
  library('sleuth')
})
```

## Associating transcripts to genes

To conform with the analyses in Boj _et al_., the walkthrough demonstrates analysis at the gene-level. Transcripts are associated to genes with the commands

```{r cache=TRUE}
mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "mmusculus_gene_ensembl",
  host = "dec2015.archive.ensembl.org")
ttg <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "transcript_version",
  "ensembl_gene_id", "external_gene_name", "description",
  "transcript_biotype"),
  mart = mart)
ttg <- dplyr::rename(ttg, target_id = ensembl_transcript_id,
  ens_gene = ensembl_gene_id, ext_gene = external_gene_name)
head(ttg)
```

## Creating the sleuth object

The next step is to prepare for differential analysis by creating a sleuth object:

```{r}
sample_ids <- dir(file.path('..', 'results'))
sample_to_condition <- dplyr::mutate(sample_to_condition,
  path = file.path('..', 'results', sample_ids, 'kallisto'))
```

```{r cache=TRUE}
so <- sleuth_prep(sample_to_condition, target_mapping = ttg,
  aggregation_column = 'ens_gene', extra_bootstrap_summary = TRUE)
```

# Basic quality control

Before beginning analysis, it is useful to examine the overall structure of the data.

```{r}
plot_pca(so, color_by = 'genotype')
```

The plot reveals a an outlier.
Additionally, because of the default legend positioning with ggplot, the image is a bit distorted.
We will fix that here and in the following plots.
The sample corresponding to the outlier can be identified by using the sample name labels in the PCA plot:

```{r}
new_position_theme <- theme(legend.position = c(0.80, 0.90))
plot_pca(so, color_by = 'genotype', text_labels = TRUE) +
  new_position_theme
```

The outlier sample is `SRR1654638`.

Removal of outliers can greatly improve results, and is sometimes warranted due to botched sample prep, problems with sequencing or occasionally book-keeping errors/sample mix-ups.
However removing outliers can also accidentally, if not intentionally, become a form of data fishing.

Therefore before removing outliers it is prudent to try to understand _why_ a sample might be an outlier. To do so it is helpful to examine the PCA loadings, i.e. the primary genes whose linear combinations define the principal components:

```{r}
plot_loadings(so, pc_input = 1)
```

The first gene driving PC1 is

```{r}
plot_bootstrap(so, 'ENSMUSG00000096887', color_by = 'genotype') +
  new_position_theme
```

The gene is highly expressed and is variable between samples, but the outlier sample does not appear to be particularly different than the other samples. However the second gene influencing PC1 reveals the source of variation causing sample SRR1654638 to be an outlier.

```{r}
plot_bootstrap(so, 'ENSMUSG00000035202', color_by = 'genotype') +
  new_position_theme
```

Sample SRR1654638 is markedly different from all of the other samples.
Looking it up at [Ensembl](http://www.ensembl.org/Mus_musculus/Gene/Summary?g=ENSMUSG00000035202;r=9:123366927-123462666) reveals that it is a mitochondrial gene.
This could be investigated further, for example by detailed analyses of alignments to mitochondrial genes; we leave the further investigation of the nature of the sample as an exercise for the reader.

The outlier can be removed with

```{r}
sample_to_condition <- dplyr::filter(sample_to_condition, sample != 'SRR1654638')
```

Once the sample has been removed it is useful to re-examine the data at a high level.
The first two principal components shown in the PCA plot now reveal a better behaved experiment and the separation between genotype is evident:


```{r cache=TRUE}
so <- sleuth_prep(sample_to_condition, target_mapping = ttg,
  aggregation_column = 'ens_gene', extra_bootstrap_summary = TRUE)
```

```{r}
plot_pca(so, color_by = 'genotype', text_labels = TRUE) +
  new_position_theme
```

It is useful to examine the data with respect to different covariates as well; coloring by sex is shown below:

```{r}
plot_pca(so, color_by = 'sex') +
  new_position_theme
```

# Testing for differential genes

The first test we perform is to identify genes that are differently expressed in either the mT or PanIN conditions.
To do this we first specify a "full" model for sleuth. This model contains parameters that are specific to both sex and condition.

To identify differential genes sleuth will compare expression values estimated according to the full model, with those of a reduced model.
In this case, the reduced model estimates parameters only according to sex.
This has the effect of modeling the data according to the assumption that expression is independent of condition.
By comparing expression estimates deirved according to the two models, sleuth is able to identify outliers that represent genes who expression can be explained much more accurately when considering condition.

The sleuth commands for performing the differential analysis are:

```{r cache=TRUE}
so <- sleuth_fit(so, ~sex + genotype, 'full')
so <- sleuth_fit(so, ~sex, 'genotype')
so <- sleuth_lrt(so, 'genotype', 'full')
full_results <- sleuth_results(so, 'genotype:full', 'lrt', show_all = FALSE)
```

A comparison with the naíve analysis as demonstrated in the [Bottomly _et al_. walkthrough](http://pachterlab.github.io/sleuth/walkthroughs/bottomly/) is intructive to perform here as well; we omit the details as they are already demonstrated in the Bottomly _et al_. walkthrough.

Instead, the walkthrough is concluded by a comparison to the anlaysis of the Boj _et al_. paper which (a) did not remove the outlier, (b) did not test conditioning on sex, and (c) performed three pairwise tests among the conditions instead of one single test (thus not properly correcting p-values for multiple testing).
The most significant genes in the sleuth analysis are:

```{r}
sleuth_significant <- dplyr::filter(full_results, qval <= 0.05)
head(sleuth_significant, 20)
```
The top gene, ENSMUSG00000099974 is Bcl2a1d. The Boj _et al_. analysis identified Bcl2a1b (see Figure their 5) as the top downregulated gene between mP vs. mN, but Bcl2a1d was not present among the top 10 genes.
The second most significant sleuth gene is Gcnt3 which has already been identified as a [target for pancreatic cancer](https://www.ncbi.nlm.nih.gov/pubmed/26880801), but is not among the top genes identified by Boj _et al_. in their Figure 5.
The third sleuth gene ENSMUSG00000030513 is PCSK6 which is thought to play a role in tumor progress (again this gene is not in Boj _et al_ Figure 5).

A detailed (re)analysis of the Boj _et al_. is beyond the scope of this walkthrough, but other (typical) next steps would be pathway enrichment analysis and follow up experiments examining top up/down regulated genes in more detail.
